pkg/skaffold/git/gitutil_test.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"errors"
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/config"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+func TestDefaultRef(t *testing.T) {
+	tests := []struct {
+		description  string
+		masterExists bool
+		mainExists   bool
+		expected     string
+		err          error
+	}{
+		{
+			description:  "master branch exists",
+			masterExists: true,
+			mainExists:   true,
+			expected:     "master",
+		},
+		{
+			description: "master branch does not exist; main branch exists",
+			mainExists:  true,
+			expected:    "main",
+		},
+		{
+			description: "master and main don't exist",
+			err:         errors.New("failed to get default branch for repo http://github.com/foo.git"),
+		},
+	}
+	for _, test := range tests {
+		testutil.Run(t, test.description, func(t *testutil.T) {
+			var f *testutil.FakeCmd
+			if test.masterExists {
+				f = testutil.CmdRunOut("git ls-remote https://github.com/foo.git master", "8be3f718c015a5fe190bebf356079a25afe0ca57  refs/heads/master")
+			} else {
+				f = testutil.CmdRunOut("git ls-remote https://github.com/foo.git master", "")
+			}
+			if test.mainExists {
+				f = f.AndRunOut("git ls-remote https://github.com/foo.git main", "8be3f718c015a5fe190bebf356079a25afe0ca58  refs/heads/main")
+			} else {
+				f = f.AndRunOut("git ls-remote https://github.com/foo.git main", "")
+			}
+			t.Override(&searchGitPath, func() (string, error) { return "git", nil })
+			t.Override(&util.DefaultExecCommand, f)
+			ref, err := defaultRef("https://github.com/foo.git")
+			t.CheckErrorAndDeepEqual(test.err != nil, err, test.expected, ref)
+		})
+	}
+}
+
+func TestSyncRepo(t *testing.T) {
+	tests := []struct {
+		description string
+		g           latest.GitInfo
+		cmds        map[string]error
+		existing    bool
+		shouldErr   bool
+		expected    string
+	}{
+		{
+			description: "first time repo clone succeeds",
+			g:           latest.GitInfo{Repo: "http://github.com/foo.git", Path: "bar/skaffold.yaml", Ref: "master"},
+			cmds: map[string]error{
```

Comment:
maps are not sorted, so these could happen in any order.

See [the helm tests](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/deploy/helm/helm_test.go#L466) for an example of how to chain examples.

Reply:
already fixed